Flaky test: TestSMTPMailerWithNoEHLO.test_send_multiple_threads

Examples of failures:

- https://travis-ci.org/zopefoundation/zope.sendmail/jobs/271167878 (pypy3)

<details><summary>log</summary>

```
$ coverage run -m zope.testrunner --test-path=src
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/home/travis/build/zopefoundation/zope.sendmail/src/zope/sendmail/tests/test_mailer.py", line 147, in run
    self.test_send_multiple_same_mailer()
  File "/home/travis/build/zopefoundation/zope.sendmail/src/zope/sendmail/tests/test_mailer.py", line 138, in test_send_multiple_same_mailer
    self.assertEqual(2, len(smtps))
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 821, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 814, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 2 != 1
Failure in test test_send_multiple_threads (zope.sendmail.tests.test_mailer.TestSMTPMailerWithNoEHLO)
Traceback (most recent call last):
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 601, in run
    testMethod()
  File "/home/travis/build/zopefoundation/zope.sendmail/src/zope/sendmail/tests/test_mailer.py", line 163, in test_send_multiple_threads
    self.assertEqual([True for _ in threads], results)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 821, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 1019, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 1001, in assertSequenceEqual
    self.fail(msg)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 666, in fail
    raise self.failureException(msg)
AssertionError: Lists differ: [True, True] != [True, AssertionError('2 != 1',)]
First differing element 1:
True
AssertionError('2 != 1',)
- [True, True]
+ [True, AssertionError('2 != 1',)]
  Ran 77 tests with 1 failures, 0 errors and 3 skipped in 2.738 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests in 0.000 seconds.
```

</details>

- https://travis-ci.org/zopefoundation/zopetoolkit/jobs/283726780 (py27)

<details><summary>log</summary>

```
Running test-ztk-zope.sendmail
test-ztk-zope.sendmail failed with:
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/opt/python/2.7.13/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/opt/python/2.7.13/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py2.7.egg/zope/sendmail/tests/test_mailer.py", line 147, in run
    self.test_send_multiple_same_mailer()
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py2.7.egg/zope/sendmail/tests/test_mailer.py", line 138, in test_send_multiple_same_mailer
    self.assertEqual(2, len(smtps))
  File "/opt/python/2.7.13/lib/python2.7/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/2.7.13/lib/python2.7/unittest/case.py", line 506, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 2 != 1
Failure in test test_send_multiple_threads (zope.sendmail.tests.test_mailer.TestSMTPMailerWithNoEHLO)
Traceback (most recent call last):
  File "/opt/python/2.7.13/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py2.7.egg/zope/sendmail/tests/test_mailer.py", line 163, in test_send_multiple_threads
    self.assertEqual([True for _ in threads], results)
  File "/opt/python/2.7.13/lib/python2.7/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/2.7.13/lib/python2.7/unittest/case.py", line 742, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
  File "/opt/python/2.7.13/lib/python2.7/unittest/case.py", line 724, in assertSequenceEqual
    self.fail(msg)
  File "/opt/python/2.7.13/lib/python2.7/unittest/case.py", line 410, in fail
    raise self.failureException(msg)
AssertionError: Lists differ: [True, True] != [True, AssertionError('2 != 1'...
First differing element 1:
True
AssertionError('2 != 1',)
- [True, True]
+ [True, AssertionError('2 != 1',)]
  Ran 77 tests with 1 failures, 0 errors and 3 skipped in 0.107 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests in 0.000 seconds.
Running test-ztk-zope.i18n
...
```

</details>

- https://travis-ci.org/zopefoundation/zopetoolkit/jobs/283726793 (pypy)

<details><summary>log</summary>

```
Running test-ztk-zope.sendmail
test-ztk-zope.sendmail failed with:
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/threading.py", line 797, in __bootstrap_inner
    self.run()
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/threading.py", line 750, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py2.7.egg/zope/sendmail/tests/test_mailer.py", line 147, in run
    self.test_send_multiple_same_mailer()
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py2.7.egg/zope/sendmail/tests/test_mailer.py", line 138, in test_send_multiple_same_mailer
    self.assertEqual(2, len(smtps))
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/unittest/case.py", line 506, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 2 != 0
Failure in test test_send_multiple_threads (zope.sendmail.tests.test_mailer.TestSMTPMailerWithNoEHLO)
Traceback (most recent call last):
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py2.7.egg/zope/sendmail/tests/test_mailer.py", line 163, in test_send_multiple_threads
    self.assertEqual([True for _ in threads], results)
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/unittest/case.py", line 742, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/unittest/case.py", line 724, in assertSequenceEqual
    self.fail(msg)
  File "/opt/python/pypy2.7-5.8.0/lib-python/2.7/unittest/case.py", line 410, in fail
    raise self.failureException(msg)
AssertionError: Lists differ: [True, True] != [True, AssertionError('2 != 0'...
First differing element 1:
True
AssertionError('2 != 0',)
- [True, True]
+ [True, AssertionError('2 != 0',)]
  Ran 77 tests with 1 failures, 0 errors and 3 skipped in 0.224 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests in 0.000 seconds.
Running test-ztk-zope.session
...
```

</details>

- https://travis-ci.org/zopefoundation/zopetoolkit/jobs/283726794 (pypy3)

<details><summary>log</summary>

```
Running test-ztk-zope.sendmail
test-ztk-zope.sendmail failed with:
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
Failure in test test_send_multiple_threads (zope.sendmail.tests.test_mailer.TestSMTPMailerWithNoEHLO)
Traceback (most recent call last):
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 601, in run
    testMethod()
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py3.5.egg/zope/sendmail/tests/test_mailer.py", line 163, in test_send_multiple_threads
    self.assertEqual([True for _ in threads], results)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 821, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 1019, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 1001, in assertSequenceEqual
    self.fail(msg)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 666, in fail
    raise self.failureException(msg)
AssertionError: Lists differ: [True, True] != [True, AssertionError('2 != 1',)]
First differing element 1:
True
AssertionError('2 != 1',)
- [True, True]
+ [True, AssertionError('2 != 1',)]
  Ran 77 tests with 1 failures, 0 errors and 3 skipped in 0.263 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests Exception in thread Thread-3:
Traceback (most recent call last):
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py3.5.egg/zope/sendmail/tests/test_mailer.py", line 147, in run
    self.test_send_multiple_same_mailer()
  File "/home/travis/build/zopefoundation/zopetoolkit/eggs/zope.sendmail-4.1.0-py3.5.egg/zope/sendmail/tests/test_mailer.py", line 138, in test_send_multiple_same_mailer
    self.assertEqual(2, len(smtps))
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 821, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/pypy3.5-5.8.0/lib-python/3/unittest/case.py", line 814, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 2 != 1
in 0.000 seconds.
Running test-ztk-zope.session
...
```

</details>

--- a/src/zope/sendmail/tests/test_mailer.py
+++ b/src/zope/sendmail/tests/test_mailer.py
@@ -15,9 +15,10 @@
 """Tests for mailers.
 """
 
+import unittest
+from functools import partial
 from ssl import SSLError
 
-import unittest
 from zope.interface.verify import verifyObject
 from zope.sendmail.interfaces import ISMTPMailer
 from zope.sendmail.mailer import SMTPMailer
@@ -91,20 +92,25 @@ class TestSMTPMailer(unittest.TestCase):
     assertRaisesRegex = getattr(unittest.TestCase, 'assertRaisesRegex',
                                 unittest.TestCase.assertRaisesRegexp)
 
-    def _makeSMTP(self, h, p):
-        self.smtp = self.SMTPClass(h, p)
-        self.smtp_hook(self.smtp)
-        return self.smtp
+    def _makeMailer(self, port=None, smtp_hook=None):
+        if port is None:
+            mailer = SMTPMailer()
+        else:
+            mailer = SMTPMailer(u'localhost', port)
+
+        def _make_smtp(host, port):
+            smtp = self.SMTPClass(host, port)
+            if smtp_hook:
+                smtp_hook(smtp)
+            return smtp
+
+        mailer.smtp = _make_smtp
+        return mailer
 
     def setUp(self, port=None):
         self.smtp = None
-        if port is None:
-            self.mailer = SMTPMailer()
-        else:
-            self.mailer = SMTPMailer(u'localhost', port)
-
-        self.mailer.smtp = self._makeSMTP
-        self.smtp_hook = lambda smtp: None
+        self.mailer = self._makeMailer(
+            port=port, smtp_hook=partial(setattr, self, 'smtp'))
 
     def test_interface(self):
         verifyObject(ISMTPMailer, self.mailer)
@@ -128,20 +134,21 @@ class TestSMTPMailer(unittest.TestCase):
         # multiple mails.
         smtps = []
 
-        def hook(smtp):
-            smtps.append(smtp)
+        # Note: this test needs to be thread-safe so it must avoid mutating
+        # attributes on `self`!
 
-        self.smtp_hook = hook
         for run in (1, 2):
             fromaddr = 'me@example.com' + str(run)
             toaddrs = ('you@example.com', 'him@example.com')
             msgtext = 'Headers: headers\n\nbodybodybody\n-- \nsig\n'
-            self.mailer.send(fromaddr, toaddrs, msgtext)
-            self.assertEqual(self.smtp.fromaddr, fromaddr)
-            self.assertEqual(self.smtp.toaddrs, toaddrs)
-            self.assertEqual(self.smtp.msgtext, msgtext)
-            self.assertTrue(self.smtp.quitted)
-            self.assertTrue(self.smtp.closed)
+            mailer = self._makeMailer(smtp_hook=smtps.append)
+            mailer.send(fromaddr, toaddrs, msgtext)
+            smtp = smtps[-1]
+            self.assertEqual(smtp.fromaddr, fromaddr)
+            self.assertEqual(smtp.toaddrs, toaddrs)
+            self.assertEqual(smtp.msgtext, msgtext)
+            self.assertTrue(smtp.quitted)
+            self.assertTrue(smtp.closed)
 
         self.assertEqual(2, len(smtps))
 
@@ -232,11 +239,12 @@ class TestSMTPMailer(unittest.TestCase):
         def hook(smtp):
             smtp.ehlo_code = 100
             smtp.helo = lambda: (100, "Nope")
-        self.smtp_hook = hook
+
+        mailer = self._makeMailer(smtp_hook=hook)
 
         with self.assertRaisesRegex(RuntimeError,
                                     "Error sending HELO to the SMTP server"):
-            self.mailer.vote(None, None, None)
+            mailer.vote(None, None, None)
 
     def test_abort_no_conn(self):
         self.assertIsNone(self.mailer.abort())

--- a/src/zope/sendmail/tests/test_mailer.py
+++ b/src/zope/sendmail/tests/test_mailer.py
@@ -133,6 +133,7 @@ class TestSMTPMailer(unittest.TestCase):
         # The mailer re-opens itself as needed when sending
         # multiple mails.
         smtps = []
+        mailer = self._makeMailer(smtp_hook=smtps.append)
 
         # Note: this test needs to be thread-safe so it must avoid mutating
         # attributes on `self`!
@@ -141,7 +142,6 @@ class TestSMTPMailer(unittest.TestCase):
             fromaddr = 'me@example.com' + str(run)
             toaddrs = ('you@example.com', 'him@example.com')
             msgtext = 'Headers: headers\n\nbodybodybody\n-- \nsig\n'
-            mailer = self._makeMailer(smtp_hook=smtps.append)
             mailer.send(fromaddr, toaddrs, msgtext)
             smtp = smtps[-1]
             self.assertEqual(smtp.fromaddr, fromaddr)

